tests/prague/eip6206_jumpf/contracts.py

```diff
@@ -0,0 +1,244 @@
+            delta = (
+                0
+                if target_non_returning or current_non_returning
+                else target_outputs - current_height
+            )
+            current_extra_push = max(0, current_height - target_height)
+            current_section = Section.Code(
+                code=Op.PUSH0 * (current_height)
+                + Op.CALLDATALOAD(0)
+                + Op.RJUMPI[1]
+                + (Op.STOP if current_non_returning else Op.RETF)
+                + Op.PUSH0 * current_extra_push
+                + Op.JUMPF[2],
+                code_inputs=0,
+                code_outputs=current_outputs,
+                max_stack_height=current_height + max(1, current_extra_push),
+            )
+            target_section = Section.Code(
+                code=((Op.PUSH0 * delta) if delta >= 0 else (Op.POP * -delta))
+                + Op.CALLF[3]
+                + (Op.STOP if target_non_returning else Op.RETF),
+                code_inputs=current_height,
+                code_outputs=target_outputs,
+                max_stack_height=max(current_height, current_height + delta),
+            )
+
+            container = Container(
+                name="target_co-%s_to-%s"
+                % (
+                    "N" if current_non_returning else current_outputs,
+                    "N" if target_non_returning else target_outputs,
+                ),
+                sections=[
+                    quick_code(Op.JUMPF[1], height=0 if current_non_returning else current_height)
+                    if current_non_returning
+                    else quick_code(
+                        Op.CALLF[1](0, 0) + Op.STOP,
+                        height=0 if current_non_returning else 2 + current_outputs,
+                    ),
+                    current_section,
+                    target_section,
+                    quick_code(Op.SSTORE(0, 1) + Op.RETF, outputs=0, height=2),
+                ],
+            )
+
+            # now sort validity...
+            if target_non_returning:
+                valid.append(container)
+            elif current_non_returning or current_outputs < target_outputs:
+                # both as non-returning handled above
+                container.validity_error = EOFException.UNDEFINED_EXCEPTION
+                invalid.append(container)
+            else:
+                # both are returning, and current >= target
+                valid.append(container)
+    return (valid, invalid)
+
+
+def generate_jumpf_stack_returning_rules():
+    """
+    Generate tests for JUMPF where we are testing the stack rules.  Returning section cases
+    """
+    valid = []
+    invalid = []
+    for current_outputs in [0, 2, 4]:
+        for target_outputs in [x for x in [0, 2, 4] if x <= current_outputs]:
+            for target_inputs in [0, 2, 4]:
+                for stack_diff in [-1, 0, 1] if target_inputs > 0 else [0, 1]:
```

Comment:
this nested loops also look like can benefit from pyspec parametrisation

Reply:
done.